### ISSUE TITLE
Extend parameterized partial and single-case active patterns to _total multi-case_ active patterns

### ISSUE THREAD

**Issue by abelbraaksma** [fsharp/fslang-suggestions#1213]
This question came up a few times recently on Slack, which suggests to me there's a real-world use case for this scenario.

I propose we extend support of parameterized active patterns to include _total_ active patterns, where the pattern has multiple choices.

There's currently no existing way of doing this in F#, such patterns are [explicitly forbidden](https://learn.microsoft.com/en-us/dotnet/fsharp/language-reference/active-patterns).

## Example

```f#
let (|Ge|Lt|) x y =
   if y >= x then Ge
   else Lt

let f() = 
    match 42 with 
    | Ge 12 -> "greater or equal"
    | Lt 12 -> "smaller"
```

## Syntax

As the above example shows, one might think that writing `Lt 11` would make the pattern match non-total. This is correct, and honestly I don't have a good way around it. Since a _total active pattern_ is considered a single block, perhaps the rule could be that only the first item should have the argument. After all, that's how such patterns are compiled currently:

**Improved example**

```f#
let (|Ge|Lt|) x y =
   if y >= x then Ge
   else Lt

let f() = 
    match 42 with 
    | Ge 12 -> "greater or equal"
    | Lt -> "smaller"
```

## Pros and Cons 

The advantages of making this adjustment to F# are
  * orthogonality between the different types of active patterns\
  * allows writing parameterized _total active patterns_

The disadvantages of making this adjustment to F# are 
  * it's work
  * some may consider the "totality check" to become ambiguous, see above for a solution.

## Extra information

Estimated cost (XS, S, M, L, XL, XXL): S

Related suggestions: I couldn't find any, though I'd be surprised if this wasn't suggested before.

## Affidavit (please submit!)

Please tick this by placing a cross in the box:
* [x] This is not a question (e.g. like one you might ask on [stackoverflow](http://stackoverflow.com)) and I have searched stackoverflow for discussions of this issue
* [x] I have [searched both open and closed suggestions on this site](http://github.com/fsharp/fslang-suggestions/issues) and believe this is not a duplicate
* [x] This is not something which has obviously "already been decided" in previous versions of F#.  If you're questioning a fundamental design decision that has obviously already been taken (e.g. "Make F# untyped") then please don't submit it.

Please tick all that apply:
* [x] This is not a breaking change to the F# language design
* [x] I or my company would be willing to help implement and/or test this

## For Readers

If you would like to see this issue implemented, please click the :+1: emoji on this issue. These counts are used to generally order the suggestions by engagement.

**Comment by kerams** [comment]
Wouldn't it be impossible to perform an exhaustiveness check?

**Comment by abelbraaksma** [comment]
Well, the example is just that, an example. Certainly there are other ways of writing that particular code. What I’m looking for is arguments to a total (exhaustive) pattern.

Maybe we should drop exhaustiveness, and allow for multi case partial patterns (which, in single case scenarios, already allow for arguments, but that leads to repetitive code and suboptimal generated code). 

Note that in the text above I do touch on a way to make the exhaustiveness check work with this sort of pattern.

**Comment by kerams** [comment]
Yeah, I suppose it could be exhaustive if it's just the first case that takes the pattern. However, then you can a DU and move the parameter up:

```fsharp
type GeLt =
    | Ge
    | Lt

let comp x y =
   if y >= x then Ge
   else Lt

let f() = 
    match comp 42 12 with 
    | Ge -> "greater or equal"
    | Lt -> "smaller"
```

That's still not an active pattern but has the same ergonomics.

> allow for multi case partial patterns

That crossed my mind too:

```fsharp
let (|G|L|_|) x y =
   if y > x then
       Some G
   elif y < x then
       Some L
   else
       None

let f() = 
    match 42 with 
    | G 500 -> "greater than 500"
    | G 200 -> "greater than 200"
    | G 50 -> "greater than 50"
    | L  2-> "less than 12"
    | _ -> "something else"
```

This can be done today with 2 active patterns. What I am trying to say is that not everything has to be solved by an active pattern if a similarly straightforward alternative exists.

**Comment by dsyme** [comment]
We always intended to do this, but the reliance on a choice type encoding held us back, plus the general complexity compared to number of use cases. The metadata allows for it in principle.

**Comment by dsyme** [comment]
I guess that qualifies as an approve in principle

**Comment by abelbraaksma** [comment]
I’ll work on an RFC. I think the main thing to decide upon will be whether a single argument should be provided (that is, only the first of the cases), or we go the more advanced route and compiler will consider same arguments in the ‘totality check’ (that is: each case with same arg translates to single execution of active pattern).

If then different args are used, unless the totality check concludes that at least once all cases with same arg were used, the normal warning for incomplete pattern matches is raised.

This (more complex) approach has my preference as it gives more freedom to consumers, but may lead to significantly more work to implement.

**Comment by abelbraaksma** [comment]
Just happened upon my own bug report again, lol. Apparently I try to use this more often than I care to admit. Hmm, time for an RFC, I guess.

Anyway, I found a clumsy workaround. Apparently, you can use active patterns with a variable in the match clause. It does require both of the `x`'s to be the same, i.e. `match x with` and `MatchLower x ...` must both be there.

This seems like a bug (or a feature?) and something to keep in mind that this might be used in the wild:

```f#
[<AutoOpen>]
module StringPatterns =
    /// Matches a string by its lowercase invariant version against the given constant string
    let (|MatchLower|_|) value matchValue =
        if String.trimToLower value = matchValue then
            Some matchValue
        else
            None

    /// Matches a string by its uppercase invariant version against the given constant string
    let (|MatchUpper|_|) value matchValue =
        if String.trimToUpper value = matchValue then
            Some matchValue
        else
            None

    let compareString x =
        match x with
        | MatchLower x "foo" -> true
        | _ -> false
```

**Comment by Wallby** [comment]
This is sort of possible like this?:

```fsharp
let (|Ge|_|) y x =
   if x >= y then Some(Ge)
   else None
let (|Lt|_|) y x =
  if x < y then Some(Lt)
  else None

let f (x:int) = 
    match x with 
    | Ge 12 -> "greater or equal"
    | Lt 12 -> "smaller"
    | _ -> null

printfn "f: %s" (f 42)
```

No way to get rid of the redundant `_ -> null` though. But passing `x` to `Ge` or `Lt` manually is not required.

**Comment by Wallby** [comment]
Or this of course works, but is harder to read? 
```fsharp
let (|Ge|Lt|) (y, x) =
   if x >= y then Ge
   else Lt

let f (x:int) = 
    match 12, x with 
    | Ge -> "greater or equal"
    | Lt -> "smaller"
```

Wouldn't this feature require some sort of evaluating that a branch of a match expression can never be reached?
So a sort of `static_assert` like feature that fails at compile time if the `_` case is reachable? Given that specifying the parameter to every case is error prone.
```fsharp
let f (x:int) = 
    match x with 
    | Ge 12 -> "greater or equal"
    | Lt 12 -> "smaller"
    | _ -> unreachable "This should be unreachable!!!"
```

**Comment by Wallby** [comment]
So this works, a non exhaustive pattern that produces a warning instead of a compilation error. Could this be implemented using `when` to get a warning, that can be suppressed using an attribute?
```fsharp
type Fraction = {
    denominator: int
    numerator: int
}

let (|IsDivisible|_|) (value: Fraction) =
    match (value.denominator, value.numerator) with
    | (denominator, numerator) when denominator % numerator = 0 -> Some()
    | (denominator, numerator) when denominator % numerator <> 0 -> None

let Test x y =
    match {denominator = x; numerator = y} with
    | IsDivisible -> true
    | _ -> false

Test 16 4
Test 16 3
```

So something like this:
```fsharp
let (|Ge|Lt|) (y, x) =
   if x >= y then Ge
   else Lt

let f (x:int) = 
    match x with 
    | when Ge 12 -> "greater or equal"
    | when Lt 12 -> "smaller"
```

**Comment by Wallby** [comment]
Ah, I figured out why the version by @abelbraaksma works. It is the same logic as in the `ParseRegex` example in the documentation: https://learn.microsoft.com/en-us/dotnet/fsharp/language-reference/active-patterns#parameterized-active-patterns

```fsharp
let (|Lt|_|) y x =
    if x < y then Some ()
    else None
let (|Ge|_|) y x =
   if x >= y then Some ()
   else None

let f (x:int) = 
    match x with
    | Ge 12 _ -> "greater or equal"
    | Lt 12 _-> "smaller"
```

> This seems like a bug (or a feature?) and something to keep in mind that this might be used in the wild:
> Funny that this is always true??

It is expected that it is always true, because `y` is a binding, not an input to MatchLower:
> ```fsharp
> let compareString x y =
>     match x with
>     | MatchLower x y -> true
>     | _ -> false
> ```

The correct version would be:
> ```fsharp
> let compareString x y =
>     match x with
>     | MatchLower y _ -> true
>     | _ -> false
> ```
